elkodon/src/config.rs

```diff
@@ -0,0 +1,320 @@
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct PublishSubscribe {
+    /// The maximum amount of supported [`crate::port::subscriber::Subscriber`]
+    pub max_subscribers: usize,
+    /// The maximum amount of supported [`crate::port::publisher::Publisher`]
+    pub max_publishers: usize,
+    /// The maximum buffer size a [`crate::port::subscriber::Subscriber`] can have
+    pub subscriber_buffer_size: usize,
+    /// The maximum amount of [`crate::sample::Sample`]s a [`crate::port::subscriber::Subscriber`] can
+    /// hold in parallel.
+    pub subscriber_max_borrowed_samples: usize,
+    /// The maximum amount of [`crate::sample_mut::SampleMut`]s a [`crate::port::publisher::Publisher`] can
+    /// loan in parallel.
+    pub publisher_max_loaned_samples: usize,
+    /// The maximum history size a [`crate::port::subscriber::Subscriber`] can request from a
+    /// [`crate::port::publisher::Publisher`].
+    pub publisher_history_size: usize,
+    /// Defines if the how the [`crate::port::subscriber::Subscriber`] buffer behaves when it is
+    /// full. When safe overflow is activated, the [`crate::port::publisher::Publisher`] will
+    /// replace the oldest [`crate::sample::Sample`] with the newest one.
+    pub enable_safe_overflow: bool,
+    /// If no safe overflow is activated it defines the deliver strategy of the
+    /// [`crate::port::publisher::Publisher`] when the [`crate::port::subscriber::Subscriber`]s
+    /// buffer is full.
+    pub unable_to_deliver_strategy: UnableToDeliverStrategy,
+}
+
+/// Default settings for the event messaging pattern. These settings are used unless
+/// the user specifies custom QoS or port settings.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Event {
+    /// The maximum amount of supported [`crate::port::listener::Listener`]
+    pub max_listeners: usize,
+    /// The maximum amount of supported [`crate::port::notifier::Notifier`]
+    pub max_notifiers: usize,
+}
+
+/// Represents the configuration elkodon will use. It is separated into two parts, the [`Global`]
+/// settings that every instance will use and the [`Defaults`] that are used unless the user does
+/// overrides them.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Config {
+    /// Global settings
+    pub global: Global,
+    /// Default settings that can be overridden by the user
+    pub defaults: Defaults,
+}
+
+static ELKODON_CONFIG: LazySingleton<Config> = LazySingleton::<Config>::new();
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            global: Global {
+                #[cfg(not(target_os = "windows"))]
+                root_path: "/tmp/elkodon/".to_string(),
+                #[cfg(target_os = "windows")]
+                root_path: "C:\\Windows\\Temp\\elkodon\\".to_string(),
+                service: Service {
+                    directory: "services".to_string(),
+                    publisher_data_segment_suffix: ".publisher_data".to_string(),
```

Comment:
Oh, what do you think of another layer of indirection. Instead of having `root_path/service::directory`  we could add another `services` directory, e.g. `root_path/services/service::directory` or maybe ``root_path/service_domains/service::directory`. Just in case there are more folders in `root_path` which are not related to services. In this case one does not need to have a list of non-service directories but can just list everything under `root_path/service_domains` to get all services.

Reply:
Dismissed, since this use case is covered by just using a different config with a different root path.